### PR TITLE
[fix JENKINS-28294] Improve the ERROR message when a Maven project is launched with a java version < to the one of the slave.

### PR DESCRIPTION
**Pull request by aheritier**

### CODE REVIEW

**Comment by aheritier** [comment]
@reviewbybees

File: src/main/java/hudson/maven/AbstractMavenProcessFactory.java [hunk 1]
```diff
@@ -302,7 +302,11 @@ public String call() throws IOException {
                 if (originalJdk == null) { // so we only try this once
                     for (Throwable t = x; t != null; t = t.getCause()) {
                         if (t instanceof UnsupportedClassVersionError || t instanceof SocketException) {
-                            listener.error("[JENKINS-18403] JDK 5 not supported to run Maven; retrying with slave Java and setting compile/test properties to point to " + jdk.getHome());
+                            listener.error("[JENKINS-18403][JENKINS-28294] JDK '" + jdk.getName() +
+                                    "' not supported to run Maven projects; retrying with slave Java and setting compile/test properties to point to " +
+                                    jdk.getHome() +
+                                    ". Maven projects have to be launched with a Java version greater or equal to the one used by the slave. " +
```
Comment (daniel-beck) [inline src/main/java/hudson/maven/AbstractMavenProcessFactory.java]: Pretty sure it's just a Maven version capable of running Jenkins. I.e. if your slave is running on Java 8, you can configure a JDK 7 here.

Comment (aheritier) [inline src/main/java/hudson/maven/AbstractMavenProcessFactory.java]: I did that test : Jenkins core 1.642.1 on JDK7 with Slave on JDK7 with Maven Job on JDK 6 launching a build with Maven 3.0.5 and it gives: 

------
02/05/16 00:14:51] [SSH] Starting sftp client.
[02/05/16 00:14:51] [SSH] Copying latest slave.jar...
[02/05/16 00:14:52] [SSH] Copied 506,256 bytes.
Expanded the channel window size to 4MB
[02/05/16 00:14:52] [SSH] Starting slave process: cd "/Users/arnaud/Temp/slave" && /Library/Java/JavaVirtualMachines/jdk1.7.0_80.jdk/Contents/Home/bin/java  -jar slave.jar
<===[JENKINS REMOTING CAPACITY]===>channel started
Slave.jar version: 2.53.2
------


------
[appaloosa] $ /Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Home/bin/java -cp /Users/arnaud/Temp/slave/maven3-agent.jar:/Users/arnaud/Temp/slave/tools/hudson.tasks.Maven_MavenInstallation/maven-3.0.5/boot/plexus-classworlds-2.4.jar org.jvnet.hudson.maven3.agent.Maven3Main /Users/arnaud/Temp/slave/tools/hudson.tasks.Maven_MavenInstallation/maven-3.0.5 /Users/arnaud/Temp/slave/slave.jar /Users/arnaud/Temp/slave/maven3-interceptor.jar /Users/arnaud/Temp/slave/maven3-interceptor-commons.jar 60564
<===[JENKINS REMOTING CAPACITY]===>channel started
ERROR: [JENKINS-18403][JENKINS-28294] JDK 'java-1.6' not supported to run Maven projects; retrying with slave Java and setting compile/test properties to point to /Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Home. Maven projects have to be launched with a JDK version greater or equal to the one used by the slave. Use the Maven JDK Toolchains (plugin) to build your maven project with an older JDK.
Established TCP socket on 60569
maven3-agent.jar already up to date
maven3-interceptor.jar already up to date
maven3-interceptor-commons.jar already up to date
[appaloosa] $ /Library/Java/JavaVirtualMachines/jdk1.7.0_80.jdk/Contents/Home/jre/bin/java -cp /Users/arnaud/Temp/slave/maven3-agent.jar:/Users/arnaud/Temp/slave/tools/hudson.tasks.Maven_MavenInstallation/maven-3.0.5/boot/plexus-classworlds-2.4.jar org.jvnet.hudson.maven3.agent.Maven3Main /Users/arnaud/Temp/slave/tools/hudson.tasks.Maven_MavenInstallation/maven-3.0.5 /Users/arnaud/Temp/slave/slave.jar /Users/arnaud/Temp/slave/maven3-interceptor.jar /Users/arnaud/Temp/slave/maven3-interceptor-commons.jar 60569
<===[JENKINS REMOTING CAPACITY]===>channel started
------

Maven 3.0.5 supports Java 6 but we can see the error when it is launched with this version. It is then started with Java 7. AFAIU the problem comes from the fact that some J7 serialised classes coming from master are executing on the J6.

Comment (aheritier) [inline src/main/java/hudson/maven/AbstractMavenProcessFactory.java]: I need to test but yes I think that Master on J7, slave on J8 and Maven on J7 should work. I need to reword my sentence.

Comment (jglick) [inline src/main/java/hudson/maven/AbstractMavenProcessFactory.java]: What Daniel said. AFAIK

**Comment by jtnord** [comment]
:bee:

**Comment by aheritier** [comment]
Thus @daniel-beck @jglick @jtnord the constraint is to launch maven jobs with a version of java >= to the minimum one supported by the master (and thus built for).

Examples:
* Jenkins >= 1.612 requires Java 7 thus Maven jobs must be launched with Java >= 7
* Jenkins >= 1.520 requires Java 6 thus Maven jobs must be launched with Java >= 6

I tested that:
* Jenkins 1.609 on J7 with a slave on J7 and a maven 3.0.5 job on J6 is OK (because 1.609 built for J6)
* Jenkins 1.625 on J7 with a slave on J7 and a maven 3.0.5 job on J6 is KO (because 1.625 built for J7)

I reworded the warning like this:

ERROR: [JENKINS-18403][JENKINS-28294] JDK 'java-1.6' not supported to run Maven projects; retrying with slave Java and setting compile/test properties to point to /Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Home. Maven projects have to be launched with a Java version greater or equal to the minimum version required by the master. Use the Maven JDK Toolchains (plugin) to build your maven project with an older JDK.

Is it better ?

**Comment by daniel-beck** [comment]
> Use the Maven JDK Toolchains (plugin) to build your maven project with an older JDK.

If we could actually provide a guide how to implement this, that would potentially very useful.

Otherwise, looks good (if we can't find out what JRE version it is -- that would be even more awesome).

**Comment by aheritier** [comment]
@daniel 

> If we could actually provide a guide how to implement this, that would potentially very useful.

Maven toolchains are documented here : https://maven.apache.org/guides/mini/guide-using-toolchains.html

I could wrote an article to explain how to use them with Jenkins. 

I think I will have to use the [Custom Tools Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Custom+Tools+Plugin) or the [Extra Tool Installers plugin](https://wiki.jenkins-ci.org/display/JENKINS/Extra+Tool+Installers+Plugin) to install some extra JDKs on the slaves (cc @oleg-nenashev @recampbell for advices) and the classical [Config File Provider plugin](https://wiki.jenkins-ci.org/display/JENKINS/Config+File+Provider+Plugin) to deploy on the slave the toolchains configuration file. But I think (cc @imod ) that there is no existing feature somewhere to generate the toolchains.xml file based on the installation of several JDKs on the slave.

where should be the best place to do this ? Wiki ?

AFAIK there is a beginning of article here : https://cloudbees.zendesk.com/hc/en-us/articles/212493238

But I could move it in the OSS side with pleasure.

> Otherwise, looks good (if we can't find out what JRE version it is -- that would be even more awesome).

Maybe I could grab more details from the UnsupportedClassVersionError. I need to check

**Comment by aheritier** [comment]
I improved a little bit the error message

```
ERROR: ================================================================================
ERROR: Invalid project setup: hudson/maven/AbstractMavenProcessFactory$ConfigureOriginalJDK : Unsupported major.minor version 51.0
ERROR: [JENKINS-18403][JENKINS-28294] JDK 'j6' not supported to run Maven projects.
ERROR: Maven projects have to be launched with a Java version greater or equal to the minimum version required by the master.
ERROR: Use the Maven JDK Toolchains (plugin) to build your maven project with an older JDK.
ERROR: Retrying with slave Java and setting compile/test properties to point to /Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Home/.
ERROR: ================================================================================
```

**Pull request merged**